src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs

```diff
@@ -65,8 +69,13 @@ private object DeserializeObject(int typeIndex)
         }
 
         // Issue https://github.com/dotnet/runtime/issues/39290 tracks finding an alternative to BinaryFormatter
-        private void InitializeBinaryFormatter()
+        private bool InitializeBinaryFormatter()
         {
+            if (!LocalAppContextSwitches.BinaryFormatterEnabled)
```

Comment:
Do we really need this? Can this depend on BinaryFormater throwing the exception instead?

Reply:
If you would like to provide an exception with a better error message, this can catch `NotSupportedException` thrown by the binary formatter.

Reply:
Perhaps letting the PNSE / NSE bubble up would be ideal after all. It clearly indicates "I know what you were trying to do, but the current app configuration disallowed it."

Reply:
We _do_ want the linker to trim all of this away when BinaryFormatter is disabled, though. That's also part of the discussion at https://github.com/dotnet/runtime/issues/39090. Maybe the best course of action should be to have this method throw an exception with the exact same error message that BinaryFormatter would have thrown.

Reply:
I think there are two options:
1. Throw the same exception as what binary formatter throws here
2. Catch the NotSupportedException from the BinaryFormatter and replace it with a different exception that has better message specific to ResourceReader.

Either way is fine with me.